Arbitrary number of skinning weights.

etherlore opened the pull request:
This change allows for an arbitrary number skinning weights per vertex. 
The number of weights allowed is controlled by the command line parameter --skinning-weights. 
This defaults to 4 so as to not change the current behavior. 
Normalization now also has a command line parameter --normalize-weights which also defaults to the current behavior of normalizing.

The entire pipeline has been upgraded to be agnostic to the number of weights per vertex. The Vec4 structures are populated at the end. Any remainder of weight/index space in the Vec4 structures are zeroed, again as the previous behavior. The code also determines the minimum set of Vec4s are globally for the mesh, and the command line parameter so as to not bloat with all zero structures. All vertices have the same number of Vec4s in the end, which I believe is what the spec requires.

zellski commented:
Thanks so much, will look this over later today.

zellski commented on src/FBX2glTF.cpp:
```diff
@@ -310,7 +323,7 @@ int main(int argc, char* argv[]) {
     raw.TransformTextures(texturesTransforms);
   }
-  raw.Condense();
+  raw.Condense(gltfOptions.maxSkinningWeights, gltfOptions.normalizeSkinningWeights);
```
Random thought: I'm starting to think we should have `RawOptions`, `FbxOptions` and `GltfOptions` interfaces, each one defined in the relevant subdirectory, then a concrete class that implements all three, and finally instantiate e.g. `RawModel` with an instance of the option class. It could be argued that it's more elegant to send to each method exactly the options it needs, but there's an equally valid point that once a method starts accepting more than 4 parameters, it's lost most of its ability to clean things up rather than muddy them.

(This method is obviously fine, just thinking outloud.)

etherlore replied:
Yes, this was actually something I was considering as well, and influenced some of the design choices in the implementation. We could have gone a different route and sent options all the way to the gltf-agnostic Fbx2Raw stuff and had the Vec4 structures set up right there. But then I think it's good to keep that part of the code agnostic to gltf limitations and structures anyway.

etherlore commented on src/gltf/Raw2Gltf.hpp:
```diff
@@ -154,6 +154,44 @@ struct AttributeDefinition {
+    glType(_glType),
+    dracoAttribute(dracoAttribute),
+    dracoComponentType(dracoComponentType),
+    arrayOffset(arrayOffset){}
```
Yes, I will do another fetch not that I should be able to get upstream thanks to your changes there.

etherlore commented on src/raw/RawModel.cpp:
```diff
@@ -614,11 +660,11 @@ void RawModel::CreateMaterialModels(
-        if ((keep & RAW_VERTEX_ATTRIBUTE_JOINT_WEIGHTS) == 0) {
-          vertex.jointWeights = defaultVertex.jointWeights;
+        if ((keep & RAW_VERTEX_ATTRIBUTE_JOINT_INDICES) == 0) {
+            vertex.jointIndices = defaultVertex.jointIndices;
```
Ok, I will take a look and make sure we're good. We switched ourselves recently on some of our projects, so that may be why it's inconsistent in some of these changes.

etherlore commented:
I updated the PR with latest, and I believe I fixed the remaining issues. I hope this was done correctly, otherwise let me know and I can do better. If the commits are too messy I could do a fresh fork and apply the changes to that. Let me know what you think!

fire commented:
@etherlore Can you review https://github.com/revolufire/FBX2glTF/commit/6875e2288e3100b29e556d9502850146f0077af3 ?

fire commented:
https://github.com/revolufire/FBX2glTF/commit/ad3a5736f3bbf3ae1aede3f1aef1cfb83113b066

I updated the title to:

> Allow arbitrary number of skinning weights.

>

> Controlled through command line parameter. Defaults to 4 skinning weights.
